### PR TITLE
Quantity operations call `__array_finalize__` too often?

### PR DESCRIPTION
**Pull request by mhvk**
This is mostly a reminder that while trying to test speed issues in #3898, we found that `Quantity.__array_finalize__` seems to be called more often than necessary. Eg.,
```
In [2]: q = np.arange(10)*u.m
In __array_finalize__; self, obj= <Quantity [ 0., 1., 2., 3., 4., 5., 6., 7., 8., 9.] (Unit not initialised)> array([ 0.,  1.,  2.,  3.,  4.,  5.,  6.,  7.,  8.,  9.])

In [3]: q[2:3]
In __array_finalize__; self, obj= <Quantity [ 2.] (Unit not initialised)> <Quantity [ 0., 1., 2., 3., 4., 5., 6., 7., 8., 9.] m>
In __array_finalize__; self, obj= <Quantity [ 2.] (Unit not initialised)> <Quantity [ 2.] m>
In __array_finalize__; self, obj= <Quantity [ 2.] m> <Quantity [ 0., 1., 2., 3., 4., 5., 6., 7., 8., 9.] m>
Out[3]: <Quantity [ 2.] m>
In [6]: np.square(q)
In __array_finalize__; self, obj= <Quantity [  6.94677988e-310,  2.34355731e-316,  6.94677779e-310,
             6.94677779e-310,  6.94677779e-310,  6.94677779e-310,
             6.94677779e-310,  6.94677779e-310,  6.94677779e-310,
             6.94677781e-310] (Unit not initialised)> array([  6.94677988e-310,   2.34355731e-316,   6.94677779e-310,
         6.94677779e-310,   6.94677779e-310,   6.94677779e-310,
         6.94677779e-310,   6.94677779e-310,   6.94677779e-310,
         6.94677781e-310])
In __array_finalize__; self, obj= <Quantity [  6.94677988e-310,  2.34355731e-316,  6.94677779e-310,
             6.94677779e-310,  6.94677779e-310,  6.94677779e-310,
             6.94677779e-310,  6.94677779e-310,  6.94677779e-310,
             6.94677781e-310] (Unit not initialised)> <Quantity [ 0., 1., 2., 3., 4., 5., 6., 7., 8., 9.] m>
```
It seems likely this can be sped up a little.

### CODE REVIEW

**Comment by mhvk** [comment]
OK, made a few simple changes that reduce the number of calls to `__array_finalize__`. In the process, also found a bug, where raising a quantity to a `Fraction` resulted in an object dtype:
```
In [4]: q = (25 * u.m**2) 

In [5]: q ** Fraction(1,2)
Out[5]: <Quantity 5.0 m>

In [6]: q2 = q ** Fraction(1,2)

In [7]: q2.dtype
Out[7]: dtype('O')
```

File: astropy/analytic_functions/blackbody.py [hunk 1]
```diff
@@ -66,7 +66,7 @@ def blackbody_nu(in_x, temperature):
     # Check if input values are physically possible
     if temp < 0:
         raise ValueError('Invalid temperature {0}'.format(temp))
-    if np.any(freq <= 0):  # pragma: no cover
+    if not np.all(np.isfinite(freq)) or np.any(freq <= 0):  # pragma: no cover
```
Comment (mhvk) [inline astropy/analytic_functions/blackbody.py]: @embray - do you know what this `pragma: no cover` line is for? This test is in fact covered. Should I remove it?

Comment (embray) [inline astropy/analytic_functions/blackbody.py]: No idea--I didn't write this code.

Comment (mhvk) [inline astropy/analytic_functions/blackbody.py]: @pllim - I think you wrote the blackbody code: do you remember what the `pragma: no cover` was for?

Comment (pllim) [inline astropy/analytic_functions/blackbody.py]: It is probably because I did not write a test for it, so I excluded it from coverage report. If there is a test now, you can remove the `pragma`.

**Comment by embray** [comment]
Seems to be good to go.  Should the Fraction bug maybe also be pulled out and backported to v1.0.5?

**Comment by embray** [comment]
Likewise it's not clear to me, now, how ea0061bef90cb5e13a486d26bd8c1426a9e883c1 relates to the original issue.

File: astropy/time/core.py [hunk 2]
```diff
@@ -794,23 +794,7 @@ def __getitem__(self, item):
...
-                    setattr(tm, attr, val[item])
-                except IndexError:  # location may be scalar (same for all)
-                    continue
-
-        # Copy other 'info' attr only if it has actually been defined.
-        if 'info' in self.__dict__:
-            tm.info = self.info
-
-        return tm
+        return self._replicate('__getitem__', item)
```
Comment (mdboom) [inline astropy/time/core.py]: Is this related to this PR?  Perhaps this commit is on this branch by accident?

**Comment by mhvk** [comment]
@mdboom - you were right that an unrelated `Time` commit had snuck in (not quite sure why). I removed it now, and think this is ready to go (assuming tests pass...)

**Comment by mdboom** [comment]
It seems like 29d4701 is also unrelated to this PR.

**Comment by mhvk** [comment]
My changes actually caused blackbody to fail, and while correcting that error, I thought I might as well add proper tests. So, maybe leave it in.

Actually, it seems the `Time.__getitem__` commit also contained something that should have remained, namely a correction to `EarthLocation`. Will fix now.

**Comment by mhvk** [comment]
@mdboom - turns out the `TIme` commit was indeed needed as is. Fortunately, it got in via #3872, so a rebase on master fixed the errors. Do think it is ready to go in now...

**Comment by mhvk** [comment]
@embray - I reassigned this to you. I think it is actually all OK, but @mdboom didn't get to do a final review.

**Comment by embray** [comment]
Okay, I've reviewed the changes (I haven't tested them out directly but I believe you that it's faster :)  Everything else seems to make sense.
